convert2rhel/backup/subscription.py

```diff
@@ -0,0 +1,137 @@
+
+        if self.enabled:
+            try:
+                subscription.unregister_system()
+            except subscription.UnregisterError as e:
+                loggerinst.warning(str(e))
+            except OSError:
+                loggerinst.warning("subscription-manager not installed, skipping")
+
+        super(RestorableSystemSubscription, self).restore()
+
+
+class RestorableAutoAttachmentSubscription(RestorableChange):
+    """
+    Auto attach subscriptions with RHSM in a fashion that can be reverted.
+    """
+
+    def __init__(self):
+        super(RestorableAutoAttachmentSubscription, self).__init__()
+        self._is_attached = False
+
+    def enable(self):
+        self._is_attached = subscription.auto_attach_subscription()
+        super(RestorableAutoAttachmentSubscription, self).enable()
+
+    def restore(self):
+        if self._is_attached:
+            subscription.remove_subscription()
+            super(RestorableAutoAttachmentSubscription, self).restore()
+
+
+class RestorableDisableRepositories(RestorableChange):
+    """
+    Gather repositories enabled on the system before we disable them and enable
+    them in the rollback.
+    """
+
+    # Look for the `Repo ID` key in the subscription-manager output, if there
+    # is a match, we save it in the named group `repo_id`. This will find all
+    # occurrences of the Repo ID in the output.
+    ENABLED_REPOS_PATTERN = re.compile(r"Repo ID:\s+(?P<repo_id>\S+)")
+
+    def __init__(self, rhel_repos_ignore):
+        super(RestorableDisableRepositories, self).__init__()
+        self._repos_to_enable = []
+        self._rhel_repos_ignore = rhel_repos_ignore or []
+
+    def _get_enabled_repositories(self):
+        """Get repositories that were enabled prior to the conversion.
+
+        :returns list[str]: List of repositories enabled prior the conversion.
+            If no repositories were enabled or match the ignored rhel
+            repositories, defaults to an empty list.
+        """
+        cmd = ["subscription-manager", "repos", "--list-enabled"]
+        output, _ = utils.run_subprocess(cmd, print_output=False)
+
+        repositories = []
+        matches = re.finditer(self.ENABLED_REPOS_PATTERN, output)
+        if matches:
+            repositories = [
+                match.group("repo_id") for match in matches if match.group("repo_id") not in self._rhel_repos_ignore
+            ]
+        return repositories
+
+    def enable(self):
+        repositories = self._get_enabled_repositories()
+
+        if repositories:
+            self._repos_to_enable = repositories
+            loggerinst.debug("Repositories enabled in the system prior to the conversion: %s" % ",".join(repositories))
+
+        subscription.disable_repos()
+        super(RestorableDisableRepositories, self).enable()
+
+    def restore(self):
+        if not self.enabled:
+            return
+
+        loggerinst.task("Rollback: Enabling RHSM repositories")
```

Comment:
```suggestion
        loggerinst.task("Rollback: Restoring state of the repositories")
```
I would say we are not enabling the RHSM repos, but the repos user had prior to the conversion

Reply:
Good point! It is not only RHSM repos. I will aply that.